### PR TITLE
options/posix: syslog improvements

### PR DESCRIPTION
**Pull request by Dennisbonke**
This PR implements `syslog()` and `vsyslog()` like musl, with the added extension that should the `connect()` call to the syslog socket fail, we fall back to using `mlibc::infoLogger()`.

### CODE REVIEW

File: options/posix/generic/syslog-stubs.cpp [hunk 1]
```diff
@@ -44,13 +66,80 @@ int setlogmask(int) {
...
 }
 
+void vsyslog(int priority, const char *message, va_list ap) {
+	int cs;
+	if(!(log_mask & LOG_MASK(priority & 7)) || (priority & ~0x3ff))
+		return;
+	//pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cs);
+	//LOCK(lock);
+	_vsyslog(priority, message, ap);
+	//UNLOCK(lock);
```
Comment (avdgrinten) [inline options/posix/generic/syslog-stubs.cpp:143]: The log can be implemented on top of `mlibc::FutexLock`, https://github.com/managarm/mlibc/blob/master/options/internal/include/mlibc/lock.hpp#L8.

File: options/posix/generic/syslog-stubs.cpp [hunk 2]
```diff
@@ -44,13 +66,80 @@ int setlogmask(int) {
...
+	va_list ap;
+	va_start(ap, format);
+	vsyslog(priority, format, ap);
+	va_end(ap);
 }
 
+void vsyslog(int priority, const char *message, va_list ap) {
+	int cs;
+	if(!(log_mask & LOG_MASK(priority & 7)) || (priority & ~0x3ff))
+		return;
```
Comment (avdgrinten) [inline options/posix/generic/syslog-stubs.cpp:139]: Log an error using infoLogger().

File: options/posix/generic/syslog-stubs.cpp [hunk 3]
```diff
@@ -44,13 +66,80 @@ int setlogmask(int) {
...
+	}
+
+	if(!(priority & LOG_FACMASK))
+		priority |= log_facility;
+
+	now = time(NULL);
+	gmtime_r(&now, &tm);
+	// HACK: %e is unsupported, replacing with a less then ideal alternative.
+	//strftime(timebuf, sizeof timebuf, "%b %e %T", &tm);
+	strftime(timebuf, sizeof timebuf, "%b %d %T", &tm);
```
Comment (avdgrinten) [inline options/posix/generic/syslog-stubs.cpp:100]: We should just implement %e.

File: options/posix/generic/syslog-stubs.cpp [hunk 4]
```diff
@@ -20,8 +32,18 @@ void closelog(void) {
...
 	log_fd = socket(AF_UNIX, SOCK_DGRAM | SOCK_CLOEXEC, 0);
-	if(log_fd >= 0)
-		connect(log_fd, (const sockaddr *)&log_addr, sizeof log_addr);
+	if(log_fd >= 0) {
+		int ret = connect(log_fd, (const sockaddr *)&log_addr, sizeof log_addr);
+		if(ret) {
+			if(errno == ENOENT) {
+				mlibc::infoLogger() << "\e[31mmlibc: connect returned ENOENT, falling back to infoLogger\e[39m" << frg::endlog;
+				use_mlibc_logger = 1;
+			} else {
```
Comment (Geertiebear) [inline options/posix/generic/syslog-stubs.cpp:41]: Why the different codepaths here? is reporting ENOENT explicitely that important?

File: options/posix/generic/syslog-stubs.cpp [hunk 5]
```diff
@@ -44,13 +66,80 @@ int setlogmask(int) {
...
+		    || send(log_fd, buf, l, 0) < 0)
+		    && (log_opt & LOG_CONS)) {
+			fd = open("/dev/console", O_WRONLY|O_NOCTTY|O_CLOEXEC);
+			if(fd >= 0) {
+				dprintf(fd, "%.*s", l - hlen, buf + hlen);
+				close(fd);
+			}
+		}
+		if(log_opt & LOG_PERROR)
+			dprintf(2, "%.*s", l - hlen, buf + hlen);
```
Comment (Geertiebear) [inline options/posix/generic/syslog-stubs.cpp:125]: Maybe don't hardcode 2 here but instead use ``STDERR_FILENO``

Comment (Dennisbonke) [reply options/posix/generic/syslog-stubs.cpp]: This has been incorporated, thanks for pointing it out.

**Comment by avdgrinten** [comment]
LGTM now.

**Pull request merged**